### PR TITLE
Add S3 SSE KMS key and bucket key encryption

### PR DESCRIPTION
**Pull request by juliusknorr**
Originally proposed by @tsdicloud this is the rebased version of https://github.com/nextcloud/server/pull/26899 with pulling the changes from https://github.com/nextmcloud/server/compare/master..s3fragment without the already merged parts from https://github.com/nextcloud/server/pull/27877

- [x] Add option to set a SSE KMS id and enable usage of bucket key encryption
- [x] Replace 'verify_bucket_exists' with 'autocreate' as documented in https://docs.nextcloud.com/server/latest/admin_manual/configuration_files/primary_storage.html#simple-storage-service-s3 which is also used in other object storage implementations as described in #26093
- [x] Double check with working patch
  - [x] https://github.com/nextmcloud/server/blob/47c00d0724ec118076ea6fcd102d62112a6a8744/lib/private/Files/ObjectStore/S3ConnectionTrait.php
  - [x] https://github.com/nextmcloud/server/blob/47c00d0724ec118076ea6fcd102d62112a6a8744/lib/private/Files/ObjectStore/S3ObjectTrait.php

### CODE REVIEW

File: lib/private/Files/ObjectStore/S3ObjectTrait.php [hunk 1]
```diff
@@ -55,7 +60,7 @@ public function readObject($urn) {
 				'Bucket' => $this->bucket,
 				'Key' => $urn,
 				'Range' => 'bytes=' . $range,
-			]);
+			] + $this->getSseKmsGetParameters());
```
Comment (artonge) [inline lib/private/Files/ObjectStore/S3ObjectTrait.php:63]: `SseKmsPutParameters` is placed under a `config` property, `SseKmsGetParameters` is not, is this right?

File: lib/private/Files/ObjectStore/S3ConnectionTrait.php [hunk 2]
```diff
@@ -95,7 +105,118 @@ public function getProxy() {
...
+		return [];
+	}
+
+	/**
+	 * Create the required bucket
+	 *
+	 * @throws \Exception if bucket creation fails
+	 */
+	protected function createNewBucket() {
+		$logger = \OC::$server->get(LoggerInterface::class);
```
Comment (artonge) [inline lib/private/Files/ObjectStore/S3ConnectionTrait.php:155]: No DI possible? :(

**Review by artonge — CHANGES_REQUESTED** [review]

File: lib/private/Files/ObjectStore/S3ConnectionTrait.php (hunk 2 above)
Comment (juliushaertl) [reply lib/private/Files/ObjectStore/S3ConnectionTrait.php:155]: No unfortunately through how the constructor is used to instantiate new instances of that class with the parameters https://github.com/nextcloud/server/blob/57db6892b74b2db1b8006257f830231deb55525b/lib/private/Files/ObjectStore/S3.php#L35

**Comment by juliushaertl** [comment]
Closing this for now as there is no current demand. This can be picked up to finish anytime of course.

**Pull request closed without merging**